src/pip/_internal/index.py

```diff
@@ -59,8 +59,50 @@
 logger = logging.getLogger(__name__)
 
 
-def _get_content_type(url, session):
-    """Get the Content-Type of the given url, using a HEAD request"""
+def _match_vcs_scheme(url):
+    """Look for VCS schemes in the URL.
+
+    Returns the matched VCS scheme, or None if there's no match.
+    """
+    from pip._internal.vcs import VcsSupport
+    for scheme in VcsSupport.schemes:
+        if url.lower().startswith(scheme) and url[len(scheme)] in '+:':
+            return scheme
+    return None
+
+
+def _is_url_like_archive(url):
+    """Check whether the URL looks like an archive.
+    """
+    filename = Link(url).filename
+    for bad_ext in ARCHIVE_EXTENSIONS:
+        if filename.endswith(bad_ext):
+            return True
+    return False
+
+
+class _NotHTML(Exception):
+    def __init__(self, content_type, response):
```

Comment:
I worry a little about attaching complex objects to exceptions. Like, could it perhaps prevent the response objects from being garbage-collected since you are later passing the exception to the logging machinery. Normally, it's best to use simple types like strings, etc, IMO. It also ensures the exceptions are serializable (e.g. for logging).

Reply:
I tend to do this to avoid assumption on what is needed at the catch side, but I guess this doesn’t really matter in this particualr scenario, since `_NotHTML` would always be internal anyway.